import_realm: Fix near-links when importing messages.

PieterCK opened the pull request:
This is stacked on top of #33590.

---

Because we rewrite the message ID and channel ID during the import,
messages containing near-links will be broken because they still point
to the old object IDs.

This PR fixes channel links, topic links, group links, dm links and
message links in the rendered content of imported messages by remapping
the relevant IDs.

Fixes #31100.
This is also a prep PR for #30166 



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK committed:
import_realm: Fix dm narrow links in exported messages.

Because we rewrite the message ID and channel ID during the import,
messages containing narrow-links will be broken because they still point
to the old object IDs.

This commit fixes dm-with links and direct message links (both private &
group dm) in the rendered content of imported messages by remapping the
relevant IDs.

Fixes #31100.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -444,6 +452,53 @@ def fix_message_rendered_content(
+        3 <= len(near_link_parts) <= 7
+        and near_link_parts[0] == "#narrow"
+        and near_link_parts[1] in ["channel", "stream"]
+    )
```
The techniques used in `is_same_server_message_link` for URL parsing would be considerably more suitable for this sort of logic.

PieterCK replied:
I ended up refactoring some logic out of `is_same_server_message_link` and consolidated it with this check, into a new function, `check_near_link_base`. This refactor is in aca6115c073375273391b176c70eadd6e7e4dc80

timabbott replied:
This seems solid.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -444,6 +452,53 @@ def fix_message_rendered_content(
+        lambda tag: tag.name == "a"
+        and tag.has_attr("href")
+        and ("#narrow/stream/" in tag.get("href") or "#narrow/channel/" in tag.get("href"))
+    )
```
You are missing some sort of check that this is for the same Zulip instance -- e.g. them being relative links (starting with `/`. I suppose also a link that's an absolute link to the pre-export server's URL should have this apply as well, but I think our Markdown system already converts all of those to relative via `rewrite_local_links_to_relative`, so we should be OK there. 

Concretely I think this means you want to be checking `startswith("/#narrow/channel/` instead.

PieterCK replied:
Ah, `rewrite_local_links_to_relative` includes the "/" as part of the prefix, so when removing it, the resulting relative link doesn't begin with "/".
 
```python
def test_relative_link(self) -> None:
        realm = get_realm("zulip")
        sender_user_profile = self.example_user("othello")
        message = Message(sender=sender_user_profile, sending_client=get_client("test"))

        msg = "http://zulip.testserver/#narrow/channel/999-hello"
        self.assertEqual(
            markdown_convert(msg, message_realm=realm, message=message).rendered_content,
            '<p><a href="#narrow/channel/999-hello">http://zulip.testserver/#narrow/channel/999-hello</a></p>',
        )

        msg = f"http://zulip.testserver/user_uploads/{realm.id}/ff/file.txt"
        self.assertEqual(
            markdown_convert(msg, message_realm=realm, message=message).rendered_content,
            f'<p><a href="user_uploads/{realm.id}/ff/file.txt">http://zulip.testserver/user_uploads/{realm.id}/ff/file.txt</a></p>',
        )
      ...
```

PieterCK replied:
I included a fix for this in this prep commit, [markdown: Fix relative link reformatting.](https://github.com/zulip/zulip/pull/32272/commits/6c723c2c183de4cc3b90763f4aff75148963bd2f)

PieterCK replied:
EDIT: this commit: 598272b27f0711bf32f0dfe2714a6bf09554fb26

timabbott replied:
Since previous Markdown renderings are in the database, I think you just need to support both `"/#narrow"` and `/#narrow` for this logic.

Meanwhile, `markdown: Fix relative link reformatting.` is a high-risk (but probably correct) change, because of how annoying it is to fix historical bugs in those links. Can you extract it to its own pull request, and post in "#api design" about it? I suspect we'll want to do your fix, but with a documented API change, and some hand-testing that mobile and terminal clients don't break.

PieterCK replied:
I opened #33907 to track this and separated the changes to this PR: #33906.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -444,6 +452,53 @@ def fix_message_rendered_content(
+        href = link["href"]
+        # This is necessary because only some near-link start
+        # with "/".
+        near_link_parts = " ".join(href.split("/")).split()
```
Hmm, I think all of them should start with `/`, see https://github.com/zulip/zulip/pull/32272/files#r1833247743

PieterCK replied:
form https://github.com/zulip/zulip/pull/32272#discussion_r1841689344:
>Ah, rewrite_local_links_to_relative includes the "/" as part of the prefix, so when removing it, the resulting relative link doesn't begin with "/".
> ...

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -444,6 +452,53 @@ def fix_message_rendered_content(
+            old_message_id = int(near_link_parts[6])
+            new_message_id = message_id_map.get(old_message_id)
+            near_link_parts[6] = str(new_message_id)
+        link["href"] = "/".join(near_link_parts)
```
This logic very much makes assumptions about the order of operators in a URL. I think that could be OK for now, but let's make the core logic for making the decision here be a pure function about URLs that lives in `zerver/lib/url_decoding.py`, so that a future sweep to harden the logic can handle that well.

Also you likely want to ensure that what is returned definitely starts with a `/`.

Further I think you have a bug for `near` links to locations in DM conversations.

PieterCK replied:
Hmm, I'm still not sure how much of this logic I should move to `url_decoding`. I've moved a good chunk of the validation and link parsing/reassembling work to c71262df91762b029257e325c414dcadfea4d191 as a helper class called `NearLinkHandler`, but the logic to remap the object IDs in the link fragment is still in `fix_near_links_in_messages` in `import_realm.py`. 

Should those remapping logic (now `remap_near_link_message_id` & `remap_near_link_recipient_encoding`) be moved to `url_decoding` / `url_encoding` too? I didn't move them because I don't think they'd be reusable elsewhere..

timabbott replied:
I think the remapping is likely not usable elsewhere, so you've probably got the right division of responsibility.

timabbott commented:
Thanks for splitting this out! I think this will be a good size of PR to iterate on in order to get a clean, correct implementation for this detail and use as a subroutine in the next steps of this project.

PieterCK commented:
Thanks for reviewing! I should have this ready for review again by tomorrow and will leave another comment then.

PieterCK commented:
Hmm, I couldn't figure out why the GitHub actions failed to install dependencies and fail the checks.

PieterCK commented:
Updates:
- 6c723c2c183de4cc3b90763f4aff75148963bd2f fixes `rewrite_local_links_to_relative` producing relative links that doesn't start with "/". See this comment https://github.com/zulip/zulip/pull/32272#discussion_r1833247743
- e83553516ca9c9375ba43fe56d6ea296c23365bf refactors `is_same_server_message_link` and added `check_near_link_base`  which contains some verification logic previously I added in `import_realm`, see this comment: https://github.com/zulip/zulip/pull/32272#discussion_r1833245490 
-  b554cb3b896657383af62e3bd3d04129966d9a0d) adds `NearLinkHandler` to help provide a more robust way of operating with near links along with some tests.
- df1c659e5c428c1051032a070f4355cbf91059b6 now can fix near links in exported private chats (group messages & direct messages)

PieterCK commented:
<details>

<summary>
git diff from the previous `import_realm.py` </summary>

```diff
git diff c2ffaa79f85f013fafd164d12ab7035c333ce8ba df1c659e5c428c1051032a070f4355cbf91059b6 -- zerver/lib/import_realm.py 
diff --git a/zerver/lib/import_realm.py b/zerver/lib/import_realm.py
index c5d3d1fcd4..020ba31b40 100644
--- a/zerver/lib/import_realm.py
+++ b/zerver/lib/import_realm.py
@@ -50,6 +50,7 @@ from zerver.lib.thumbnail import THUMBNAIL_ACCEPT_IMAGE_TYPES, BadImageError, ma
 from zerver.lib.timestamp import datetime_to_timestamp
 from zerver.lib.upload import ensure_avatar_image, sanitize_name, upload_backend, upload_emoji_image
 from zerver.lib.upload.s3 import get_bucket
+from zerver.lib.url_decoding import NearLinkHandler
 from zerver.lib.user_counts import realm_user_count_by_role
 from zerver.lib.user_groups import create_system_user_groups_for_realm
 from zerver.lib.user_message import UserMessageLite, bulk_insert_ums
@@ -463,50 +464,62 @@ def fix_message_rendered_content(
             )
 
 
-def valid_near_link_parts(near_link_parts: list[str]) -> bool:
-    return (
-        3 <= len(near_link_parts) <= 7
-        and near_link_parts[0] == "#narrow"
-        and near_link_parts[1] in ["channel", "stream"]
-    )
+def remap_near_link_recipient_encoding(fragments: list[str]) -> None:
+    section, recipient_encoding = fragments[1], fragments[2]
+    old_recipient_id, recipient_name = recipient_encoding.split("-", maxsplit=1)
+
+    id_map = {"channel": ID_MAP["stream"], "dm": ID_MAP["user_profile"]}.get(section, {})
+
+    old_id_list = map(int, old_recipient_id.split(","))
+    new_id_list = [str(id_map.get(old_id, old_id)) for old_id in old_id_list]
+
+    new_recipient_id = ",".join(new_id_list)
+    fragments[2] = f"{new_recipient_id}-{recipient_name}"
+
+
+def remap_near_link_message_id(fragments: list[str]) -> None:
+    message_id_map = ID_MAP["message"]
+    old_id = fragments[-1]
+    new_id = message_id_map.get(int(old_id), "dds")
+    fragments[-1] = str(new_id)
 
 
 def fix_near_links_in_messages(rendered_content: str) -> str:
     soup = BeautifulSoup(rendered_content, "html.parser")
+    near_link_prefixes = ["/#narrow/channel", "/#narrow/stream", "/#narrow/dm"]
+
+    # Include legacy links without the leading "/", which were
+    # used in older versions.
+    near_link_prefixes += [prefix.removeprefix("/") for prefix in near_link_prefixes]
+
     near_links = soup.find_all(
         lambda tag: tag.name == "a"
         and tag.has_attr("href")
-        and ("#narrow/stream/" in tag.get("href") or "#narrow/channel/" in tag.get("href"))
+        and (tag.get("href").startswith(tuple(near_link_prefixes)))
     )
 
     if near_links == []:
         return rendered_content
 
-    message_id_map = ID_MAP["message"]
-    channel_id_map = ID_MAP["stream"]
     for link in near_links:
-        href = link["href"]
-        # This is necessary because only some near-link start
-        # with "/".
-        near_link_parts = " ".join(href.split("/")).split()
-
-        if not valid_near_link_parts(near_link_parts):
-            continue
-        if near_link_parts[1] == "stream":
-            near_link_parts[1] = "channel"
-        encoded_channel_parts = near_link_parts[2].split("-")
-        old_channel_id = int(encoded_channel_parts[0])
-        new_channel_id = channel_id_map.get(old_channel_id)
-        if new_channel_id is None:
+        url = link["href"]
+        try:
+            near_link_instance = NearLinkHandler(url)
+        except AssertionError:
+            # NearLinkHandler does additional checks to make sure
+            # the URL is a near link. In this case it's probably
+            # not a near link (or a faulty one).
             continue
-        encoded_channel_parts[0] = str(new_channel_id)
-        near_link_parts[2] = "-".join(encoded_channel_parts)
-
-        if len(near_link_parts) == 7:
-            old_message_id = int(near_link_parts[6])
-            new_message_id = message_id_map.get(old_message_id)
-            near_link_parts[6] = str(new_message_id)
-        link["href"] = "/".join(near_link_parts)
+
+        fragments = near_link_instance.get_near_link_fragment_parts()
+        remap_near_link_recipient_encoding(fragments)
+
+        if fragments[-2] == "near" and len(fragments) >= 5:
+            remap_near_link_message_id(fragments)
+
+        near_link_instance.patch_near_link_fragment_parts(fragments)
+        link["href"] = near_link_instance.get_url()
+
     return str(soup)

```
</details>

PieterCK commented:
@prakhar1144 This should be ready for you to take a look! let me know if any of the changes are unclear

PieterCK commented on zerver/lib/import_realm.py:
```diff
@@ -455,6 +464,66 @@ def fix_message_rendered_content(
 
+def remap_near_link_recipient_encoding(fragments: list[str]) -> None:
+    section, recipient_encoding = fragments[1], fragments[2]
+    old_recipient_id, recipient_name = recipient_encoding.split("-", maxsplit=1)
```
this `split` is called with `maxsplit=1` if the channel name contains spaces, those will be replaced with "-"

PieterCK commented:
Update: did some touch-ups on code comments, variable names, and commit description.

timabbott commented on zerver/lib/url_decoding.py:
```diff
@@ -3,24 +3,66 @@
+
+    return (
+        category == "narrow"
+        and section in {"channel", "stream", "dm"}
```
I think `pm-with` is supposed to be here as an old alias for `dm` too, if this is going to be correct. Let's make sure the test suite covers both that and `stream` here.

timabbott commented on zerver/lib/url_decoding.py:
```diff
@@ -66,3 +66,56 @@ def is_same_server_message_link(url: str) -> bool:
+        fragments = "/".join(fragment_parts)
+        changed_parts["fragment"] = fragments
+        cleaned_split_result = split_result._replace(**changed_parts)
+        return cleaned_split_result
```
I think it's particularly visible here, but we have the wrong data structures for here, and it results in the parsing being wrong in corner cases that no known code produces, but should be valid. 

I think what you want to be doing is parsing the URL into a `list[NarrowParameter]` structure, and then look up the `channel` term using functions like this one in `filter.ts`.

```
    operands(operator: string): string[] {                                                              
        return this._terms                                                                              
            .filter((term) => !term.negated && term.operator === operator)                              
            .map((term) => term.operand);                                                               
    }                                                                                                   
```

It might make sense to just do a prep PR doing the necessary parsing infrastructure, and then rework this to use that. It could be:

- First commit: Rename `NarrowTerm` in `zerver/lib/narrow_helpers.py` to `NeverNegatedNarrowTerm`, which can be a subclass of a `NarrowTerm` class that has the missing `negated` flag.
- Add a  `parse_narrow_url` in `url_decoding.py` that parses an arbitrary narrow URL into a `List[NarrowTerm]`, broadly following the approach `hash_util.parse_narrow`, with tests.

And then we can merge that, and this can then be implemented on top of that helper -- and would probably involve writing something a bit like a super-simplified form of `class Filter` in the web app to support these cleaning operations but with nice abstractions for "get me the channel operator, if any", but without trying to access any data (like checking if the stream ID is valid, etc.). The overall goal here would be to have as little of the more complex logic as possible interact with a URL with all the messy `/` splitting as possible -- we'll just generate a `List[NarrowTerm]`, use it to compute a new one that's modernized and has updated IDs, and then turn that `List[NarrowTerm]` into a new URL.

(A reason this is important, beyond just safety/correctness if someone writes a `/#narrow/topic/10/channel/11/near/16` URL, which by the spec is supposed to work just like if the operators weren't in a funny order, is that we're going to need to extend what this PR does with `near` links to support `with/{message_id}` too, and at some point you just want an actual parser into nice data structures)

PieterCK replied:
I opened #33116 to add `parse_narrow_url`. sorry for the delay! I haven't been working on this until now partly because I wasn't familiar with the adjacent code base and mechanism (narrow) yet :sweat_smile: 

Excited to continue this project! but let me know if it'd be better to wait until after release : )

PieterCK replied:
Ok, I'll prioritize this

PieterCK committed:
url_decoding: Add the base `Filter` class.

This is the Python copy of a class of the same name in frontend's
`filter.Filter` class. Although it roughly follows the same set
validations, it probably won't have the same set of methods with the
frontend's `Filter`. It does not validate the referenced objects(users
and channels), and validates terms differently--e.g, converts operand to
prefered types when possible--, adjusting for its use in the backend.

The server version will primarily be used to fix narrow links in
exported messages during import.

PieterCK committed:
url_decoding: Add `Filter.operands` method.

The Python copy of a class method with the same name in frontend's
`filter.Filter`.

PieterCK committed:
url_decoding: Add `Filter.fix_redundant_is_private` method.

The Python copy of a class method with the same name in frontend's
`filter.Filter`.

PieterCK committed:
url_decoding: Add `Filter.sorted_term_types` method.

The Python copy of a class method with the same name in frontend's
`filter.Filter`. Key differences are:

- The `sorted_term_types` static method in the Python version does not
do local-aware sort in the case where two term types are the same. In
most cases we expect the list of term types to be unique.

- The `sorted_term_types` class method (not the static method) is omited
because Python does not support conflicting function names. We also
don't want to accidentally use stale term types in later operation to
update narrow links in messages during import.

PieterCK committed:
url_decoding: Add `Filter.generate_message_url` method.

This adds a method to build a message URL from the NarrowTerms in a
Filter instance. It can build channel message or dm message URLs
according to the combination of NarrowTerms given.

This could be used in the operation to fix narrow links in exported
messages during import.

PieterCK committed:
narrow_helper: Create a `NarrowTermOperandT` type alias.

This is a prep commit to add the `Filter.operands` method which could
use the same type alias.

PieterCK committed:
url_decoding: Add `Filter.update_term` method.

This adds a new method to update a specific term in a Filter instace.
This could be used in the operation to fix narrow links in
exported messages during import.

PieterCK committed:
url_decoding: Add `Filter._check_either_channel_or_dm_narrow`.

PieterCK committed:
url_decoding: Add `Filter.generate_channel_url` method.

PieterCK committed:
url_decoding: Add `Filter.generate_topic_url` method.

PieterCK committed:
url_decoding: Add `Filter.generate_dm_with_url` method.

PieterCK committed:
url_decoding: Add `Filter.get_terms` method.

This adds a utility method to get specific NarrowTerms from a Filter
instance. This could be used in the operation to fix narrow links in
exported messages during import.

PieterCK committed:
url_decoding: Refactor a `is_same_server_narrow_link`.

This refactors out the logic that checks whether a narrow link is from
the same Zulip server or not from `is_same_server_message_link`.

This is done so that it can be used to check a more general narrow link.

PieterCK committed:
import_realm: Fix channel narrow links in exported messages.

Because we rewrite the message ID and channel ID during the import,
messages containing narrow-links will be broken because they still point
to the old object IDs.

This fixes channel link, topic link and channel message narrow links.

PieterCK commented on zerver/lib/import_realm.py:
```diff
@@ -496,6 +510,166 @@ def fix_message_rendered_content(
+        if fixed_narrow_url := fix_channel_narrow_url(
+            channel_narrow_url, ID_MAP["stream"], ID_MAP["message"], realm
+        ):
+            link["href"] = fixed_narrow_url
```
This returns `None` at the slightest sign of error because this is probably far from being the most important operation when importing data, in fact, we should probably strictly keep this from raising any error.

Should we add extra guard rails here?

```python
try:
      if fixed_narrow_url := fix_channel_narrow_url(
          channel_narrow_url, ID_MAP["stream"], ID_MAP["message"], realm
      ):
          link["href"] = fixed_narrow_url
except:
       continue
````

PieterCK replied:
Do we want to return the user to home "#" in the case where we fail to update the narrow link? Currently we left them as is

PieterCK committed:
url_encoding: Add `get_message_narrow_link`.

This is adds a modern version of `near_pm_message_url`. It uses the
`url_decoding.Filter` class so that the narrow terms processing and
validation logic is centralized there.

Some modules are imported lazily to avoid circular-import.

PieterCK commented:
@zulipbot remove "has conflict"

PieterCK commented:
@zulipbot remove "has conflicts"

PieterCK commented on zerver/lib/import_realm.py:
```diff
@@ -492,6 +507,252 @@ def fix_message_rendered_content(
+        narrow_link: str = link["href"]
+
+        if narrow_link.startswith(CHANNEL_NARROW_PREFIXES) and (
+            updated_channel_narrow_url := fix_channel_narrow_url(
```
This and `fix_direct_message_narrow_url`  returns `None` at the first sign of error because we probably don't want to stop the import process over a failure from this whole operation. 

Should we add extra guard rails here? something like:

```python
  try:
      if narrow_link.startswith(CHANNEL_NARROW_PREFIXES) and (
            updated_channel_narrow_url := fix_channel_narrow_url(
                narrow_link,
                ID_MAP["stream"],
                ID_MAP["message"],
                realm,
            )
        ):
  except:
       continue
````

PieterCK commented on zerver/lib/import_realm.py:
```diff
@@ -492,6 +507,252 @@ def fix_message_rendered_content(
+    try:
+        return terms_handler.generate_message_url()
+    except InvalidOperatorCombinationError:
+        return None
```
Do we want to update the link to point elsewhere such as home ("#") for any of the cases where the operation fails? currently this leaves those links as is

PieterCK replied:
One argument to deliberately point them elsewhere is to avoid stale narrow links from accidentally pointing to valid objects in the imported realm.
